File: Introduction to Data Structures: Array and String/1. Introduction to Array/Find Pivot Index/FindPivotIndex.py
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        # helper method to calculate rightSum
        def calcRightSum(totalSum, leftSum, currentVal):
            return totalSum - leftSum - currentVal

        # calculate the sum of nums
        totalSum = sum(nums)

        # create a variable to keep track of left sum
        leftSum = 0

        # iterate throught the list
        for i in range(0, len(nums)):

            # if left sum = right sum
            if i != 0:
                leftSum += nums[i-1]

            if leftSum == calcRightSum(totalSum, leftSum, nums[i]):

                # return current index
                return i

        # return -1
        return -1

File: Introduction to Data Structures: Array and String/1. Introduction to Array/Find Pivot Index/FindPivotIndex.java
class Solution {
    // create helper method to calculate right sum 
    public static int calculateRightSum(int totalSum, int leftSum, int currentVal) {
        return totalSum - leftSum - currentVal; 
    }

    // create helper method to calcuate total sum
    public static int calculateTotalSum(int [] nums) {
        int total = 0;

        for (int i = 0; i < nums.length; i++) {
            total = total + nums[i];
        }

        return total;
    }
    
    public int pivotIndex(int[] nums) {
        
        // calculate the total sum of the array 
        int totalSum = calculateTotalSum(nums);
        
        // create variable for the left sum
        int leftSum = 0;
        
        // iterate through the array
        for (int i = 0; i < nums.length; i++) {
            // System.out.println("i: " + i);
            // System.out.println("leftSum: " + leftSum);
            // System.out.println("rightSum: " + calculateRightSum(totalSum, leftSum, nums[i]));
        
            // if the index is not zero
            if (i != 0) {
                // add current value to left sum
                leftSum = leftSum + nums[i-1];
            }
            
            // if left sum == right sum
            if (leftSum == calculateRightSum(totalSum, leftSum, nums[i])) {
                // return index
                return i;
            }
        }
        return - 1;
        
        
    }
}